### PR TITLE
Start implementing bootstore: Replicated storage required for rack unlock

### PR DESCRIPTION
**Pull request by andrewjstone**
Currently RSS is in charge of generating the rack secret and the bootstrap agent is in charge of establishing sprockets sessions and transferring shares required for rack unlock. The shares themselves, and the trust quorum membership is stored in a file on each sled. The major problems with this implementation are that it is not resilient to failure and does not allow reconfiguration of the trust quorum. Additionally, the trust quorum code is somewhat scattered across the sled-agent, and a little hard to follow.

This change introduces the `bootstore`, which will become responsible for all trust quorum protocol and storage on the rack. The bootstore will be instantiated by the boot agent, which will be responsible for establishing sprockets sessions that the trust quorum can operate over. This allows for a readable, testable implementation of the protocol and storage without having to worry about extraneous details related to networking or the SP. We may choose to pull that in later to a higher level module in the bootstore. I intend to utilize property based tests with fault injection to cover the protocol.

This is an early WIP of the bootstore, and lays out the basic structure of the code, while implementing some small bits. The goal is to get this out early before it gets too large to easily review. 

The `Node` is the main entry point to the codebase, although most logic has not been implemented. The `Node` uses the `Db` to query and store trust quorum information. The goal is to build up a fairly general two-phase commit mechanism that can also be used for storing networking information required to setup NTP so that CockroachDb can boot and the rest of the control plane software can run. The two-phase commit is standard, but as always has the problem of any node failing halts the protocol. We will manage this by moving to a new epoch and reconfiguration if this occurs. Of course the devil is in the details, but it's not incredibly complicated. This implementation will supercede what is written in section 5 of RFD 238. I will update that document with the new protocol. One thing that was not considered there was transferring the encrypted (wrapped) root secret to new nodes that don't have the original root secret derived from the epoch 0 rack secret. As part of this, we'll distribute it as part of the 2-phase commit. There are a couple mechanisms with various levels of security, that I won't get into here. The important thing to know for this PR is that I created a table and model for storing the secret. Code to use it is unimplemented so far.

There is also the notion of a [`Coordinator`] that is not yet implemented. The coordinator interacts with the sled-agent to record important information in nexus and get instructions related to trust quorum from nexus. It only stores state in memory and can crash without consequence. It may end up being driven by a saga.

I also copied the `trust_quorum` directory from the sled-agent as the `bootstore` will become responsible for the trust quorum. I haven't yet removed this code from the sled-agent, as this work is preliminary and I don't want to hook it in yet.

### CODE REVIEW

**Comment by andrewjstone** [comment]
@jgallagher One more thing I should mention with regards to this PR. There's no locking or concurrency support of any kind on the DB connection yet. That's because I haven't figured out what I want to do. I want to get more code written first. In the majority of cases, the DB will be Read-Only, and I'm not sure if we want to just cache state at the app layer and use a single connection with an `Arc<Mutex<SqliteConnection>>`, do this without the app layer caching, use multiple connections, possibly one per Node created on demand, a connection pool, or something else. In short, don't worry about concurrency on this particular PR. It'll get decided upon and implemented shortly in a follow up.

File: bootstore/src/db/models.rs [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+#[derive(Debug, Queryable, Insertable)]
+pub struct KeySharePrepare {
+    pub epoch: i32,
+    pub share: Share,
+    pub share_digest: Sha3_256Digest,
+}
+
+#[derive(Debug, Queryable, Insertable)]
+pub struct KeyShareCommit {
```
Comment (andrewjstone) [inline bootstore/src/db/models.rs:27]: I'm wondering if this table should just be a boolean `committed` field in `KeySharePrepares`. That would probably simplify queries a bit.

File: bootstore/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,18 @@
+//! The two communication paths for the bootstore:
```
Comment (jgallagher) [inline bootstore/src/lib.rs:1]: Nit: Should add the MPL header to this file

File: bootstore/src/messages.rs [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    /// Epoch 0 is the only epoch in which prepared data can be overwritten.
+    /// This is because there is no global datastore (CockroachDB) with which
+    /// to persist information that needs to be distributed to all sleds in the
+    /// rack. If RSS dies, or more importantly, the scrimlet it is running on
+    /// dies, we want to enable re-running on a different scrimlet. Since we
+    /// don't know what information was already transferred to sleds, due to
+    /// lack of global datastore, we must re-issue the request.
+    ///
+    /// This request generates a `KeySharePrepare` for epoch 0. Once all sleds
+    /// have prepared, RSS trigger the start of CockroachDB replicas. The trust
```
Comment (jgallagher) [inline bootstore/src/messages.rs:44]:
```suggestion
    /// have prepared, RSS triggers the start of CockroachDB replicas. The trust
```

File: bootstore/src/db/mod.rs [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+        &mut self,
+        epoch: i32,
+        share: SerializableShareDistribution,
+    ) -> Result<(), Error> {
+        use schema::key_share_prepares::dsl;
+        // We save the digest so we don't have to deserialize and recompute most of the time.
+        // We'd only want to do that for a consistency check occasionally.
+        let val = serde_json::to_string(&share)?;
+        let share_digest =
+            sprockets_common::Sha3_256Digest(Sha3_256::digest(&val).into())
```
Comment (jgallagher) [inline bootstore/src/db/mod.rs:98]: I think this is fine but just want to think out loud - we're hashing the `serde_json`'ified `share` here, but then inserting `share` directly into the `KeySharePrepare` that is then committed to the DB. Is there any chance we're sensitive to minor changes in serde_json that wouldn't affect correctness from a purely JSON point of view (reordering fields, adding or removing whitespace) by splitting things this way?

~~This might become clear from other parts of the PR I haven't read yet, but how does diesel serialize the `share` when inserting into the DB?~~ (Answered this question when reading `json_new_type`)

 I'm wondering if to check the digest, someone has to deserialize `share` then reserialize it with `serde_json` and hope that whatever version of `serde_json` they're using serialized it byte-for-byte identically to the version used to compute this digest. I think that is not necessary, right? The string we insert into the `share` column should be byte-for-byte identical to `val` in this function, assuming `serde_json::to_string()` is deterministic (which I think it should be for simple structs).

File: bootstore/src/db/mod.rs [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+            .collect()
+    }
+
+    fn rand_db_name() -> String {
+        let seed: String = thread_rng()
+            .sample_iter(&Alphanumeric)
+            .take(16)
+            .map(char::from)
+            .collect();
+        format!("/tmp/testdb-{}.sqlite", seed)
```
Comment (jgallagher) [inline bootstore/src/db/mod.rs:171]: Could we use `:memory:` instead for tests, to not have to worry about temp files? If not, do we need to wrap this is something that will delete it when we're done? (Maybe using `tempfile`?)

File: bootstore/src/db/models.rs (hunk 1 above)
Comment (jgallagher) [reply bootstore/src/db/models.rs:27]: Seems reasonable to me; otherwise we have the epoch/share_digest duplicated and queries have to look for both (I assume that's what you were implying with simplifying queries?). At that point does the table name become just `KeyShares`?

File: bootstore/src/node.rs [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+
+use crate::db::Db;
+use crate::messages::*;
+use crate::trust_quorum::SerializableShareDistribution;
+
+/// An error returned by a Node
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error("Version {0} messages are unsupported.")]
+    UnsupportedVersion(u32),
```
Comment (jgallagher) [inline bootstore/src/node.rs:26]: Why isn't this a case in `NodeError` instead?

**Review by jgallagher — COMMENTED** [review]
LGTM; I appreciate splitting this out into an initial PR before it becomes larger - thanks!

File: bootstore/src/db/models.rs (hunk 1 above)
Comment (andrewjstone) [reply bootstore/src/db/models.rs:27]: Yeah, that's my reasoning. And yeah, I think it just becomes `KeyShares`.

File: bootstore/src/db/mod.rs (hunk 4 above)
Comment (andrewjstone) [reply bootstore/src/db/mod.rs:98]: That's a good question. The JSON will be serialized once when inserted into CockroachDB and once here. If serialization was not deterministic, we would have a problem. In fact, what we really want is "canonical" serialization, but I don't believe serde_json provides that. This is required by TUF as well fwiw, and other crypto projects. From what I can tell though this isn't a problem for us, unless the serialization format drastically changes, which while unlikely is a bit scare. We have the need to hash and sign all this data and much other data, so maybe JSON isn't the right choice for a format here.

File: bootstore/src/db/mod.rs (hunk 5 above)
Comment (andrewjstone) [reply bootstore/src/db/mod.rs:171]: I specifically didn't worry about cleanup because:
1. These DBs are going to be quite small.
2. We probably want to be able to inspect them on test failures.

1 is just a lazy punt on my part and not a justification. 2 is potentially important, but maybe we can hook into cargo test somehow and output a file only on failure?

In any case, for now I will convert to :memory: as you suggest. That makes the most sense in the short term.

File: bootstore/src/node.rs (hunk 6 above)
Comment (andrewjstone) [reply bootstore/src/node.rs:26]: My initial thinking was that printable errors like this would end up getting converted to `NodeErrors` which themselves are part of a `NodeOpResult`, which itself is part of a `NodeResponse`. But that is problematic because we then lose the correlation `id`. 

Really, the underlying issue is that I want to be able to return errors with the `?` operator from the handle methods. But I can do that by just returning a `NodeError`. The top level code just has to know how package that up into an eventual `NodeResponse`.  Maybe `handle` should just return a `NodeResponse` no matter what. The only downside there is that tests have to match to get the error out of the response, but that seems reasonable? I can also just create a `dispatch` method called by handle that returns a `result`, and then all `handle` does is package that into a `NodeResponse`. That seems like potentially overkill though.

Comment (jgallagher) [reply bootstore/src/node.rs]: What if you put a full-fledged `Result` into `NodeResponse`? Something like

```rust
pub struct NodeResponse {
    pub version: u32,
    pub id: u64,
    pub result: Result<NodeOpResult, NodeError>,
}
```

and then remove the `Error` case from `NodeOpResult`?

Comment (andrewjstone) [reply bootstore/src/node.rs]: I changed the code to return `Result<NodeOpResult, NodeError>` from the handler methods, and changed `handle` itself to just return `NodeResponse`. This works, its just a touch weird because we end up packaging a `NodeError` into `NodeOpResult` inside `handle`. But all in all I prefer it to having every method have to do that packaging and not be able to use the `?` operator.

File: bootstore/src/db/mod.rs (hunk 5 above)
Comment (jgallagher) [reply bootstore/src/db/mod.rs:171]: I think there are checks in CI that successful tests do not leave leftovers behind in `/tmp`. If you want to keep them around for failures, the simplest thing might be to use something like https://docs.rs/tempfile/latest/tempfile/struct.Builder.html#method.tempdir to create a temp directory that will be deleted on drop, and put the file in there.

File: bootstore/src/node.rs (hunk 6 above)
Comment (andrewjstone) [reply bootstore/src/node.rs]: Fixed in [07e043e](https://github.com/oxidecomputer/omicron/pull/1656/commits/07e043e547b44d9b299d3014407ea1f85ad457b0)

Thanks!

File: bootstore/src/db/mod.rs (hunk 5 above)
Comment (andrewjstone) [reply bootstore/src/db/mod.rs]: Fixed to use `:memory:` in https://github.com/oxidecomputer/omicron/pull/1656/commits/a9ac17c3df90c34267b77f05f6b1a1730658d76c

File: bootstore/src/db/models.rs (hunk 1 above)
Comment (andrewjstone) [reply bootstore/src/db/models.rs]: Fixed in https://github.com/oxidecomputer/omicron/pull/1656/commits/10058ed9b702f4d78501d99bf443189a29fbb1dd

File: bootstore/src/lib.rs (hunk 2 above)
Comment (andrewjstone) [reply bootstore/src/lib.rs:5]: Done.

File: bootstore/src/db/mod.rs (hunk 4 above)
Comment (andrewjstone) [reply bootstore/src/db/mod.rs:86]: Ok, I used [BCS](https://github.com/diem/bcs) instead of `serde_json` as of https://github.com/oxidecomputer/omicron/pull/1656/commits/815f8e16ad77412fc23f1be6f257100912de1ad6

It's the only serde backed canonical serialization format I could find and seems perfectly suitable here. It's practically a drop in replacement.

**Comment by davepacheco** [comment]
I wanted to add some notes on the pq-sys stuff since we spent some time digging into it.  To summarize:

- we have a new crate
- it wants a dev dependency on omicron-common
- it thus depends on pq-sys (via diesel)
- [it must thus use the omicron-rpaths pattern](https://github.com/oxidecomputer/omicron/blob/067c678a21ff5983ecac8201770e82088809206e/rpaths/src/lib.rs#L5-L33), which involves a build dependency on omicron-rpaths and a regular dependency on pq-sys version "*"

We observed this behavior:

- if we have a dev dependency on pq-sys, and if the build.rs runs unconditionally, then you get a compile-time error about DEP_PQ_LIBDIRS not being set.  This happens because this variable is set by Cargo when that variable is provided by a _non-dev_ dependency that emits the variable from its build script.  pq-sys emits this.  But since we're using it as a dev dependency in this case, it's not there in our build script.
- if you have a dev dependency on pq-sys, and if the build.rs is wrapped with `#[cfg(test)]`, then you get a runtime error from ld.so.1 about not finding libpq.  You might think this solution would do the right thing by causing non-test binaries to get no RPATH entries (since the build script wouldn't have done anything) and test binaries to get the right RPATH entries (since the build script would use omicron-rpaths in that case).  But we concluded that [build scripts are run only once for the whole package](https://doc.rust-lang.org/cargo/reference/build-scripts.html#life-cycle-of-a-build-script), not once per crate (binary) built by the package.  It presumably gets run without `#[cfg(test)]`.  So you just wind up building a binary that depends on pq-sys (because it's a dev-dependency and you're building a unit test) but that's missing the RPATH entry.  Hence this failure mode.
- If we have a regular dependency on pq-sys, it all works.

Because build scripts appear to only run once and also only get metadata from regular deps (not dev deps), the conclusion is that if your package ever depends on pq-sys directly or indirectly, even via dev-dependency, then you must apply the pattern by adding a *regular* dependency on pq-sys.  You can't use a dev-dependency for pq-sys, even if you only depend on pq-sys via a dev-dependency.
